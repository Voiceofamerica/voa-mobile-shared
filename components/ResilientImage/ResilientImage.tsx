
import * as React from 'react'
import { Subscription } from 'rxjs/Subscription'

import { waitUntilOnline } from 'helpers/resilience'

import { resilientImage, imageContent } from './ResilientImage.scss'

export interface Props {
  src: string
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imageStaus: ImageStatus
}

const makeAttemptUrl = (src: string) => `${src}#${Date.now()}`
const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  componentWillMount () {
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  tryFetchImage = () => {
    if (!navigator.onLine) {
      return waitUntilOnline().then(() => this.tryFetchImage())
    }

    const { onLoadDone = noop } = this.props

    const img = new Image()

    img.addEventListener('error', () => {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    })

    img.addEventListener('load', () => {
      this.setImageStatus('loaded')
      onLoadDone()
    })

    img.src = this.props.src
  }

  setImageStatus = (imageStaus: ImageStatus) =>
    this.setState({
      imageStaus,
    })

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus } = this.state
    const { src, defaultSrc = require('./imagedefault.gif') } = this.props

    const srcToUse = imageStaus === 'loaded'
                   ? src
                   : defaultSrc

    return (<img src={srcToUse} className={imageContent} />)
  }

  render () {
    const { shouldRender = navigator.onLine } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style}>
          { this.renderContent() }
        </div>
      )
    }
  }
}

export default ReilientImage
