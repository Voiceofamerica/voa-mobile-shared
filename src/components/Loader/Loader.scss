
@import "../../commonStyles/index";

.loader {
  position: relative;
  display: flex;
  flex-direction: column;
  flex: 1;
  justify-content: center;
  align-items: center;
  font-size: 10vw;
  color: $voa-blue;

  .reloadButton {
    background: $light-grey;
    border: none;
    outline: none;
    border-radius: 5px;
    padding: 5px 30px;
    font-size: 0.5em;
  }

  .backdrop {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: -1;

    .fader {
      height: 100%;
      width: 100%;
      background: rgba(255, 255, 255, 0.3);
    }
  }

  .spinner {
    width: 25vw;
    height: 25vw;
  }
}
