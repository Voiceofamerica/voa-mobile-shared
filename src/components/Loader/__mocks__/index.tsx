
import * as React from 'react'

export default (props) => (
  React.createElement('loader', { props }, props.children)
)
