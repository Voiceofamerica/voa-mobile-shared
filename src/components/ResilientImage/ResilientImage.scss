
@import '../../commonStyles/index.scss';

.resilientImage {
  position: relative;
  background: black;

  .imageContent {
    object-fit: cover;
    width: 100%;
    height: 100%;

    &.containImage {
      object-fit: contain;
      height: auto;
      max-height: 100%;
    }
  }

  .childrenArea {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .spinner {
    color: $voa-blue;
    height: 80%;
    max-height: 100px;
    margin-top: -1%;
  }
}
