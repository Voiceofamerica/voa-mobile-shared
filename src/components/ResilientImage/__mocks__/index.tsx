
import * as React from 'react'

export default (props) => (
  React.createElement('resilient-image', { props }, props.children)
)
