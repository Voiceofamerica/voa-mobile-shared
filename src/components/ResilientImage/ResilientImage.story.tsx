
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import ResilientImage from './ResilientImage'

const imageUrl = 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png'

storiesOf('ResilientImage', module)
  .add('display', () => (
    <ResilientImage src={imageUrl} />
  ))
  .add('display without image', () => (
    <ResilientImage src='asdf' />
  ))
