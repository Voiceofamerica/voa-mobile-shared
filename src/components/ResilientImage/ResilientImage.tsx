
import * as React from 'react'

import { waitUntilOnline } from '../../helpers/resilience'

import Spinner from '../Spinner'

import { resilientImage, imageContent, childrenArea, containImage, spinner } from './ResilientImage.scss'

export interface Props {
  src: string | null
  showSpinner?: boolean
  className?: string
  alwaysShow?: boolean
  style?: React.CSSProperties
  defaultSrc?: string
  onLoadDone?: () => void
  onClick?: () => void
  contain?: boolean
}

export type ImageStatus = 'loading' | 'loaded' | 'error'

export interface State {
  shouldRender?: boolean
  imgUrl?: string
  imageStaus: ImageStatus
}

const noop = () => null
const IMAGE_FETCH_RETRY_RATE = 2000

class ReilientImage extends React.Component<Props, State> {
  state: State = {
    imageStaus: 'loading',
  }

  mounted: boolean = false

  readonly abortController = new AbortController()

  componentWillMount () {
    this.mounted = true
    const shouldRender = navigator.onLine || this.props.alwaysShow
    this.setState({
      shouldRender,
    })

    if (shouldRender) {
      this.tryFetchImage()
    }
  }

  componentWillUnmount () {
    this.mounted = false
    this.abortController.abort()
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.src !== this.props.src) {
      this.setImageStatus('loading')

      const shouldRender = navigator.onLine || this.props.alwaysShow
      this.setState({
        shouldRender,
      })

      if (shouldRender) {
        this.tryFetchImage()
          .catch(err => {
            console.warn(`Something went wrong loading ${nextProps.src}`, err)
          })
      }
    }
  }

  tryFetchImage = async () => {
    const { src } = this.props

    if (!src || !this.mounted) {
      return
    }

    if (!navigator.onLine) {
      waitUntilOnline().then(() => this.mounted && this.tryFetchImage())
      return
    }

    const { onLoadDone = noop } = this.props

    try {
      const blobUrl = await fetch(src, {
          method: 'GET',
          signal: this.abortController.signal,
        })
        .then(res => res.blob())
        .then(blob => URL.createObjectURL(blob))

      this.setImageStatus('loaded', blobUrl)
      onLoadDone()

    } catch (err) {
      this.setImageStatus('error')
      setTimeout(this.tryFetchImage, IMAGE_FETCH_RETRY_RATE)
    }
  }

  setImageStatus = (imageStaus: ImageStatus, imgUrl?: string) => {
    if (!this.mounted) {
      return
    }

    this.setState({
      imageStaus,
      imgUrl,
    })
  }

  renderOffline = () => (
    <div style={{ width: 0, height: 0 }} />
  )

  renderContent = () => {
    const { imageStaus, imgUrl } = this.state
    const { defaultSrc = require('./imagedefault.gif'), contain } = this.props

    const isLoaded = imageStaus === 'loaded'
    const srcToUse = isLoaded && imgUrl
                   ? imgUrl
                   : defaultSrc

    const className = contain ? `${imageContent} ${containImage}` : imageContent

    return (<img src={srcToUse} className={className} />)
  }

  renderSpinner = () => {
    const { showSpinner } = this.props
    const { imageStaus } = this.state

    const isLoaded = imageStaus === 'loaded'

    if (isLoaded || !showSpinner) {
      return null
    }

    return (
      <div className={childrenArea}>
        <Spinner className={spinner} />
      </div>
    )
  }

  renderChildren = () => {
    const { children } = this.props

    return (
      <div className={childrenArea}>
        {children}
      </div>
    )
  }

  render () {
    const { shouldRender = navigator.onLine || this.props.alwaysShow } = this.state
    if (!shouldRender) {
      return this.renderOffline()
    } else {
      const {
        className = '',
        style,
        onClick = () => null,
      } = this.props

      return (
        <div className={`${resilientImage} ${className}`} style={style} onClick={ev => onClick()}>
          { this.renderContent() }
          { this.renderSpinner() }
          { this.props.children ? this.renderChildren() : null }
        </div>
      )
    }
  }
}

export default ReilientImage
