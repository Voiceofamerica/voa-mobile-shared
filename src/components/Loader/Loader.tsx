
import * as React from 'react'

import Spinner from '../Spinner'
import ResilientImage from '../ResilientImage'

import { loader, backdrop, fader, reloadButton } from './Loader.scss'

export interface Props extends React.Props<HTMLDivElement> {
  loading: boolean
  error: any
  refetch: () => void
  networkStatus: number
  errorText: string | JSX.Element
  retryText: string | JSX.Element
  backgroundImage: string
  hasContent?: boolean
  className?: string
  style?: React.CSSProperties
}

export interface State {
  forceLoader: boolean
}

export default class Loader extends React.PureComponent<Props, State> {
  state: State = {
    forceLoader: false,
  }

  componentWillReceiveProps (newProps: Props) {
    if (newProps.networkStatus !== this.props.networkStatus) {
      this.toggleForceLoader(false)
    }
  }

  toggleForceLoader = (forceLoader: boolean = !this.state.forceLoader) =>
    this.setState({ forceLoader })

  render () {
    const { loading, error, errorText, retryText, backgroundImage, children, className = '', style, hasContent = false } = this.props

    const fullClassName = `${loader} ${className}`

    if (loading && !hasContent || this.state.forceLoader) {
      return (
        <div className={fullClassName} style={style}>
          <ResilientImage className={backdrop} src={backgroundImage} defaultSrc={backgroundImage}>
            <div className={fader} />
          </ResilientImage>
          <Spinner style={{ height: '20vw', width: '20vw' }} />
        </div>
      )
    } else if (error && !hasContent) {
      return (
        <div className={fullClassName} style={style}>
          {errorText}
          <button className={reloadButton} onClick={this.retry}>{retryText}</button>
        </div>
      )
    } else {
      return children as JSX.Element
    }
  }

  private retry = () => {
    this.props.refetch()
    this.toggleForceLoader(true)
  }
}
