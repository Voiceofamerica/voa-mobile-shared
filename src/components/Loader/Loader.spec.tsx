
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import ResilientImage from '../ResilientImage'
import { reloadButton } from './Loader.scss'
import Loader, { State } from './Loader'

jest.mock('../ResilientImage')
jest.mock('../Spinner')

describe('<Loader />', () => {
  describe('shapshots', () => {
    it('should render only children without an error and not loading', () => {
      const element = create((
        <Loader
          loading={false}
          error={null}
          networkStatus={1}
          refetch={jest.fn()}
          errorText='An Error Occurred'
          retryText='Retry'
          backgroundImage='some.png'
        >
          This should show up
        </Loader>
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render the spinner and background image when loading', () => {
      const element = create((
        <Loader
          loading={true}
          error={null}
          networkStatus={1}
          refetch={jest.fn()}
          errorText='An Error Occurred'
          retryText='Retry'
          backgroundImage='some.png'
          >
            This should NOT show up
          </Loader>
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render the error text and retry button when an error has occurred', () => {
      const element = create((
        <Loader
          loading={false}
          error='This is an error!'
          networkStatus={1}
          refetch={jest.fn()}
          errorText='An Error Occurred'
          retryText='Retry'
          backgroundImage='some.png'
          >
            This should NOT show up
          </Loader>
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('refetch', () => {
    it('should be called whenever the retry button is pressed', () => {
      const refetch = jest.fn()
      const element = shallow((
        <Loader
          loading={false}
          error='This is an error!'
          networkStatus={1}
          refetch={refetch}
          errorText='An Error Occurred'
          retryText='Retry'
          backgroundImage='some.png'
          >
            Content
          </Loader>
      ))

      element.find(`.${reloadButton}`).simulate('click')
      expect(refetch).toHaveBeenCalledTimes(1)
    })
    it('should force the loader to be shown', () => {
      const refetch = jest.fn()
      const element = shallow((
        <Loader
          loading={false}
          error='This is an error!'
          networkStatus={1}
          refetch={refetch}
          errorText='An Error Occurred'
          retryText='Retry'
          backgroundImage='some.png'
          >
            Content
          </Loader>
      ))

      let state: State = element.state()
      expect(state.forceLoader).toBe(false)
      expect(element.find(ResilientImage)).toHaveLength(0)

      element.find(`.${reloadButton}`).simulate('click')
      state = element.state()
      expect(state.forceLoader).toBe(true)
      expect(element.find(ResilientImage)).toHaveLength(1)
    })
  })
})
