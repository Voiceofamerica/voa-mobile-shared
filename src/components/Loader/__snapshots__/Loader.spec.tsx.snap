// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> shapshots should render only children without an error and not loading 1`] = `"This should show up"`;

exports[`<Loader /> shapshots should render the error text and retry button when an error has occurred 1`] = `
<div
  className="loader "
  style={undefined}
>
  An Error Occurred
  <button
    className="reloadButton"
    onClick={[Function]}
  >
    Retry
  </button>
</div>
`;

exports[`<Loader /> shapshots should render the spinner and background image when loading 1`] = `
<div
  className="loader "
  style={undefined}
>
  <resilient-image
    props={
      Object {
        "children": <div
          className="fader"
        />,
        "className": "backdrop",
        "defaultSrc": "some.png",
        "src": "some.png",
      }
    }
  >
    <div
      className="fader"
    />
  </resilient-image>
</div>
`;
